### PR TITLE
Use Bazel transitions

### PR DESCRIPTION
**Pull request by RamaneekGill**
@hchauvin can you provide a summary of what you did and why this was needed?

### CODE REVIEW

**Comment by hchauvin** [comment]
The gist of the problem is that when you build a docker image on Mac OSX, with Bazel you need to do a cross-platform build.  This is especially relevant with `nodejs_image`, as without a cross-platform build the Mac OSX node binary is taken to create a container layer.  This gives an "exec format error" when running the container, as the executable is expected to be a Linux binary, not a Mac OSX binary.  Therefore, to properly use `nodejs_image` on Mac OSX, it is important to cross-build Linux artifacts.

There are two ways of doing this:
- Option 1: Use the `--platforms` command-line option when invoking Bazel.
- Option 1: Use configuration transitions to force a Linux cross-build when gathering the Bazel dependencies for a Docker image.

Option 1 does not require changing `rules_docker`, but it means that the docker containers must be built separately and cannot be used as dependencies to rules running on Mac OSX, such as end-to-end test rules.  This greatly limits what can be done in terms of end-to-end testing during local development.

Option 2 requires changing `rules_docker` so that it uses [configuration transitions](https://docs.bazel.build/versions/master/skylark/lib/transition.html).  Configuration transitions were introduced precisely for this purpose and are even referenced on an issue concerning `go_image` (https://github.com/bazelbuild/rules_docker/issues/690).

I must say that this is something I had been looking for some time ago, as otherwise working with docker images on Mac OSX is too awkward.  For my projects, when dealing with docker containers, I actually had to switch to a Linux box or use remote execution.  This PR solves the issue.

**Comment by hchauvin** [comment]
My two cents as to why this has not yet happened on `rules_docker`: Googlers use Linux desktops.

File: lang/image.bzl [hunk 1]
```diff
@@ -34,8 +34,8 @@ def _binary_name(ctx):
     # /app/foo/bar/baz/blah
     return "/".join([
         ctx.attr.directory,
-        ctx.attr.binary.label.package,
-        ctx.attr.binary.label.name,
+        ctx.attr.binary[0].label.package,
```
Comment (RamaneekGill) [inline lang/image.bzl:37]: Why does this change to only specify the first element/object in `ctx.attr.binary` need to be made?

File: lang/image.bzl [hunk 2]
```diff
@@ -148,7 +148,7 @@ def _app_layer_impl(ctx, runfiles = None, emptyfiles = None):
     parent_parts = _get_layers(ctx, ctx.attr.name, ctx.attr.base)
     filepath = _final_file_path if ctx.attr.binary else layer_file_path
     emptyfilepath = _final_emptyfile_path if ctx.attr.binary else _layer_emptyfile_path
-    dep = ctx.attr.dep or ctx.attr.binary
+    dep = (ctx.attr.dep or ctx.attr.binary)[0]
```
Comment (RamaneekGill) [inline lang/image.bzl:151]: It seems like indexing at `0` will also be applied to `ctx.attr.dep`. Why did this change need to be accounted for `dep` object as well as the `binary` object?

File: lang/image.bzl [hunk 3]
```diff
@@ -263,11 +275,14 @@ image = struct(
         # The dependency whose runfiles we're appending.
         # If not specified, then the layer will be treated as the top layer,
         # and all remaining deps of "binary" will be added under runfiles.
```
Comment (RamaneekGill) [inline lang/image.bzl:277]: Do these comments need to be updated with our new change?

File: lang/image.bzl [hunk 4]
```diff
@@ -263,11 +275,14 @@ image = struct(
...
-        "dep": attr.label(),
+        "dep": attr.label(cfg = container_transition),
         "directory": attr.string(default = "/app"),
         "entrypoint": attr.string_list(default = []),
         "legacy_run_behavior": attr.bool(default = False),
         "workdir": attr.string(default = ""),
+        "_whitelist_function_transition": attr.label(
```
Comment (RamaneekGill) [inline lang/image.bzl:283]:
```suggestion
        "_allowlist_function_transition": attr.label(
```

❤️  https://www.theregister.co.uk/2019/09/03/chromium_microsoft_offensive/

The change also helps non-native English speakers comprehend this code.

File: tools/whitelists/function_transition_whitelist/BUILD [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package_group(
+    name = "function_transition_whitelist",
```
Comment (RamaneekGill) [inline tools/whitelists/function_transition_whitelist/BUILD:2]:
```suggestion
    name = "function_transition_allowlist",
```

File: lang/image.bzl [hunk 6]
```diff
@@ -263,11 +275,14 @@ image = struct(
...
+            default = "//tools/whitelists/function_transition_whitelist",
```
Comment (RamaneekGill) [inline lang/image.bzl:284]:
```suggestion
            default = "//tools/whitelists/function_transition_allowlist",
```

File: tools/whitelists/function_transition_whitelist/BUILD (hunk 5 above)
Comment (RamaneekGill) [reply tools/whitelists/function_transition_whitelist/BUILD:2]: Be sure to change the filename as well.

**Review by RamaneekGill — COMMENTED** [review]
Couple comments for you @hchauvin

**Comment by RamaneekGill** [comment]
> My two cents as to why this has not yet happened on `rules_docker`: Googlers use Linux desktops.

Macbooks that ssh into a Linux machine ;)

That and potentially being forced to use Chromebooks...

File: lang/image.bzl (hunk 1 above)
Comment (hchauvin) [reply lang/image.bzl:37]: When `cfg` is set to a transition configuration, we have what is called an "Outgoing Edge Transition" (https://docs.bazel.build/versions/master/skylark/config.html#outgoing-edge-transitions).  Outgoing edge transitions are 1:1 or 1:2+: you can basically say "I want this dependency both for Linux, for iOS, and for Mac OSX".  So the attribute is sneakily changed to an array.  The order of what you get in this array is defined by the transition.

File: lang/image.bzl (hunk 2 above)
Comment (hchauvin) [reply lang/image.bzl:151]: Because both dep and binary now use a transition configuration.

File: lang/image.bzl (hunk 3 above)
Comment (hchauvin) [reply lang/image.bzl:277]: I don't think so.  Using a transition should actually be the user's expectation.  It's the current behavior that is weird on Mac OSX.

File: lang/image.bzl (hunk 4 above)
Comment (hchauvin) [reply lang/image.bzl:283]: Unfortunately... This name is hardcoded in Bazel, I cannot change it.  This stuff is not even documented.  Even the path "tools/whitelists/function_transition_whitelist" is hardcoded.

File: tools/whitelists/function_transition_whitelist/BUILD (hunk 5 above)
Comment (hchauvin) [reply tools/whitelists/function_transition_whitelist/BUILD:2]: Hardcoded

File: lang/image.bzl (hunk 6 above)
Comment (hchauvin) [reply lang/image.bzl:284]: Massively hardcoded

**Comment by hchauvin** [comment]
I mean when I say massively hardcoded... This is what I'm talking about: https://github.com/bazelbuild/bazel/blob/a3421e204f22e478382bb8487b96df9dcd8e463a/src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkRuleClassFunctions.java#L736

I cannot change anything about this attribute at all.

**Comment by hchauvin** [comment]
One thing I'm not sure about is whether people use Bazel to build Windows containers on Windows (I'm not talking about Linux containers on Windows, but proper Windows containers). I'm not even  sure it is possible with the current `rules_docker`, so it's better to ask around before overcomplicating the code to detect the OS.

**Pull request closed without merging**